src/Features/Core/Portable/Diagnostics/Analyzers/RemoveUnnecessaryCastDiagnosticAnalyzerBase.cs

```diff
@@ -3,76 +3,70 @@
-            DiagnosticCategory.Style,
-            DiagnosticSeverity.Hidden,
-            isEnabledByDefault: true,
-            customTags: DiagnosticCustomTags.Unnecessary);
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
 
-        #region Interface methods
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptor);
-        public bool OpenFileOnly(Workspace workspace) => false;
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, this.SyntaxKindsOfInterest);
 
-        public sealed override void Initialize(AnalysisContext context)
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
         {
-            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
-            context.EnableConcurrentExecution();
+            var diagnostic = TryRemoveCastExpression(
+                context.SemanticModel,
+                (TCastExpression)context.Node,
+                context.CancellationToken);
 
-            context.RegisterSyntaxNodeAction(
-                nodeContext =>
-                    {
-                        if (TryRemoveCastExpression(nodeContext.SemanticModel, nodeContext.Node, out var diagnostic, nodeContext.CancellationToken))
-                        {
-                            nodeContext.ReportDiagnostic(diagnostic);
-                        }
-                    },
-                this.SyntaxKindsOfInterest.ToArray());
+            if (diagnostic != null)
+            {
+                context.ReportDiagnostic(diagnostic);
+            }
         }
 
-        public abstract ImmutableArray<TLanguageKindEnum> SyntaxKindsOfInterest { get; }
-
-        #endregion
-
-        protected abstract bool IsUnnecessaryCast(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken);
-        protected abstract TextSpan GetDiagnosticSpan(SyntaxNode node);
-
-        private bool TryRemoveCastExpression(
-            SemanticModel model, SyntaxNode node, out Diagnostic diagnostic, CancellationToken cancellationToken)
+        private Diagnostic TryRemoveCastExpression(SemanticModel model, TCastExpression node, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            diagnostic = default;
-
             if (!IsUnnecessaryCast(model, node, cancellationToken))
             {
-                return false;
+                return null;
             }
 
             var tree = model.SyntaxTree;
-            var span = GetDiagnosticSpan(node);
-            if (tree.OverlapsHiddenPosition(span, cancellationToken))
+            if (tree.OverlapsHiddenPosition(node.Span, cancellationToken))
```

Comment:
this seems kind of thing system should takes care, not each analyzer? @dotnet/roslyn-analysis

Reply:
That would probably be nice.  Basically, each fixer today generally has to check for this, and we're hugely inconsistent.  Having this be the default behavior (or a flag we could set once) would be nice.

Reply:
+1